### PR TITLE
Draft: [DSv32] Add triton fusion for _get_logits_head_gate

### PR DESCRIPTION
**Pull request by trevor-m**
## Motivation

Optimize DSv3.2 decode

## Modifications

Added a triton kernel to fuse 2 unary mul + 1 binary mul in `_get_logits_head_gate`.

## Accuracy Tests

Added unit test.

End-to-end:
```
Accuracy: 0.957
Invalid: 0.000
Latency: 82.141 s
Output throughput: 1558.434 token/s
```

## Profiling

Original: 6.683us
Fused: 1.479us

Screenshot 2025-10-09 at 2 29 57 PM

Screenshot 2025-10-09 at 2 31 50 PM

## End to end Benchmarking
For some reason it appears to be worse in the end-to-end benchmarks.

Before
```
============ Serving Benchmark Result ============
Successful requests:                     256       
Benchmark duration (s):                  110.47    
Total input tokens:                      236166    
Total generated tokens:                  234891    
Request throughput (req/s):              2.32      
Output token throughput (tok/s):         2126.19   
Total Token throughput (tok/s):          4263.93   
---------------Time to First Token----------------
Mean TTFT (ms):                          47615.35  
Median TTFT (ms):                        47699.76  
P99 TTFT (ms):                           76635.09  
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          64.91     
Median TPOT (ms):                        64.60     
P99 TPOT (ms):                           106.38    
---------------Inter-token Latency----------------
Mean ITL (ms):                           64.76     
Median ITL (ms):                         33.03     
P99 ITL (ms):                            38.39     
==================================================
```
After
```
============ Serving Benchmark Result ============
Successful requests:                     256       
Benchmark duration (s):                  121.11    
Total input tokens:                      236166    
Total generated tokens:                  234891    
Request throughput (req/s):              2.11      
Output token throughput (tok/s):         1939.46   
Total Token throughput (tok/s):          3889.45   
---------------Time to First Token----------------
Mean TTFT (ms):                          48331.25  
Median TTFT (ms):                        47919.13  
P99 TTFT (ms):                           81273.82  
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          75.92     
Median TPOT (ms):                        76.09     
P99 TPOT (ms):                           118.81    
---------------Inter-token Latency----------------
Mean ITL (ms):                           75.71     
Median ITL (ms):                         39.82     
P99 ITL (ms):                            45.60     
==================================================
```

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).

### CODE REVIEW

**Comment by Fridge003** [comment]
Hi @trevor-m , can you post the result with GPQA benchmark with this command:
```bash
python3 -m sglang.test.run_eval --port 30000 --eval-name gpqa --num-examples 198 --max-tokens 4096 --repeat 10
```

File: python/sglang/srt/layers/attention/nsa/fused_logits_head_gate.py [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+    assert weights.shape[1] == q_scale.shape[1], "Number of heads mismatch"
+    assert weights.is_contiguous(), "weights must be contiguous"
+
+    M, N = weights.shape
+
+    output = torch.empty((M, N, 1), dtype=torch.float32, device=weights.device)
+
+    n_heads_inv_sqrt = n_heads**-0.5
+
+    BLOCK_SIZE_M = min(32, triton.next_power_of_2(M))
```
Comment (Fridge003) [inline python/sglang/srt/layers/attention/nsa/fused_logits_head_gate.py:109]: Is there any reason of setting `BLOCK_SIZE_M` to 32?
Cauz I feel `M` might be a large number, then the number of blocks can be too large

**Comment by yuan-luo** [comment]
It seems the perf result dropped. Or you pasted in wrong order?

**Comment by trevor-m** [comment]
> Hi @trevor-m , can you post the result with GPQA benchmark with this command:
> 
> ```shell
> python3 -m sglang.test.run_eval --port 30000 --eval-name gpqa --num-examples 198 --max-tokens 4096 --repeat 10
> ```

@Fridge003 

```
====================
Repeat: 10, mean: 0.758
Scores: ['0.763', '0.778', '0.732', '0.773', '0.753', '0.732', '0.758', '0.732', '0.763', '0.798']
====================
{'chars': np.float64(3142.247474747475), 'chars:std': np.float64(1869.2599422460062), 'score:std': np.float64(0.4015072103909452), 'score': np.float64(0.797979797979798)}
Total latency: 256.655 s
Score: 0.798
```

**Comment by trevor-m** [comment]
> It seems the perf result dropped. Or you pasted in wrong order?

@yuan-luo Yeah strangely the end-to-end perf got worse. I'm currently looking into it.

File: python/sglang/srt/layers/attention/nsa/fused_logits_head_gate.py (hunk 1 above)
Comment (trevor-m) [reply python/sglang/srt/layers/attention/nsa/fused_logits_head_gate.py:109]: No reason, just copied what some other triton kernels used. Do you have any suggestions?

**Comment by trevor-m** [comment]
@Fridge003 I found that torch.compile gives good end to end speedup (1.8%). Since it's much simpler, I think I can close this PR and open one for torch.compile instead - does that sound good?

With torch.compile
```
============ Serving Benchmark Result ============
Successful requests:                     256       
Benchmark duration (s):                  50.31     
Total input tokens:                      236041    
Total generated tokens:                  235141    
Request throughput (req/s):              5.09      
Output token throughput (tok/s):         4673.86   
Total Token throughput (tok/s):          9365.60   
---------------Time to First Token----------------
Mean TTFT (ms):                          5657.55   
Median TTFT (ms):                        6100.96   
P99 TTFT (ms):                           8742.13   
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          44.53     
Median TPOT (ms):                        44.24     
P99 TPOT (ms):                           49.73     
---------------Inter-token Latency----------------
Mean ITL (ms):                           44.51     
Median ITL (ms):                         41.25     
P99 ITL (ms):                            47.54     
==================================================
```
Without
```
============ Serving Benchmark Result ============
Successful requests:                     256       
Benchmark duration (s):                  51.19     
Total input tokens:                      236041    
Total generated tokens:                  235141    
Request throughput (req/s):              5.00      
Output token throughput (tok/s):         4593.42   
Total Token throughput (tok/s):          9204.41   
---------------Time to First Token----------------
Mean TTFT (ms):                          5673.64   
Median TTFT (ms):                        5751.11   
P99 TTFT (ms):                           9266.61   
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          45.47     
Median TPOT (ms):                        45.36     
P99 TPOT (ms):                           50.81     
---------------Inter-token Latency----------------
Mean ITL (ms):                           45.44     
Median ITL (ms):                         41.65     
P99 ITL (ms):                            47.81     
==================================================
```

**Comment by Fridge003** [comment]
@trevor-m  Yeah, torch compile can be used it works better

**Comment by trevor-m** [comment]
Closing - opened https://github.com/sgl-project/sglang/pull/11565 for torch compile instead
